src/parseAndDecorate.ts

```diff
@@ -0,0 +1,219 @@
+	let decorationsMap = new Map<vscode.TextEditorDecorationType, vscode.Range[]>([
+		[argDecoration, []],
+		[selectDecoration, []],
+		[pluralDecoration, []],
+	]);
+	// Clear diagnostics
+	diagnostics.set(editor.document.uri, []);
+	let diagnosticsList: vscode.Diagnostic[] = [];
+
+	// Map of arguments for each message key
+	let placeHoldersForKey = new Map<String, Literal[]>();
+
+	// Only trigger on arb files
+	if (!editor || !path.basename(editor.document.fileName).endsWith('.arb')) {
+		return;
+	}
+	let nestingLevel = 0;
+	let placeholderLevel: number | null;
+	let metadataLevel: number | null;
+	let messageKey: string | null;
+	let definedPlaceholders: string[] = [];
+	visit(editor.document.getText(), {
+		onLiteralValue: (value: string, offset: number) => {
+			if (nestingLevel === 1) {
+				try {
+					decorateMessage(value, offset, decorationsMap, editor, true);
+				} catch (error: any) {
+					if (String(error).startsWith('Error: Unbalanced ')) {//Very hacky, but better than not checking at all... The error has no special type, unfortunately.
+						showErrorAt(offset + 1, offset + value.length + 1, 'Unbalanced curly bracket found. Try escaping the bracket using a single quote \' .', vscode.DiagnosticSeverity.Error);
+					} else {
+						throw error;
+					}
+				}
+			}
+		},
+		onObjectBegin: (offset: number, length: number, startLine: number, startCharacter: number, pathSupplier: () => JSONPath) => {
+			nestingLevel++;
+		},
+		onObjectProperty: (property: string, offset: number, length: number, startLine: number, startCharacter: number, pathSupplier: () => JSONPath) => {
+			if (placeholderLevel === nestingLevel - 1) {
+				if (!placeHoldersForKey.get(messageKey!)!.some((literal: Literal, index: number, array: Literal[]) => literal.value === property)) {
+					showErrorAt(offset + 1, offset + property.length + 1, 'Placeholder is being declared, but not used in message.', vscode.DiagnosticSeverity.Warning);
+				}
+				definedPlaceholders.push(property);
+				decorateAt(offset + 1, offset + property.length + 1, argDecoration);
+			}
+			if (nestingLevel === 1) {
+				const isMetadata = property.startsWith('@');
+				const propertyOffsetEnd = offset + property.length + 1;
+				const propertyOffsetStart = offset + 1;
+				if (isMetadata) {
+					const isGlobalMetadata = property.startsWith('@@');
+					const messageKeyExists = placeHoldersForKey.has(property.substring(1));
+					if (!isGlobalMetadata && !messageKeyExists) {
+						showErrorAt(propertyOffsetStart, propertyOffsetEnd, `Metadata for an undefined key. Add a message key with the name ${property.substring(1)}.`, vscode.DiagnosticSeverity.Error);
+					}
+					metadataLevel = nestingLevel;
+				} else {
+					if (keyNameRegex.exec(property) !== null) {
+						messageKey = property;
+						placeHoldersForKey.set(messageKey, []);
+					} else {
+						showErrorAt(propertyOffsetStart, propertyOffsetEnd, 'This is not a valid message key.', vscode.DiagnosticSeverity.Error);
```

Comment:
It may be useful to include the invalid values in each of these diagnostic messages like:

> 'foo' is not a valid message key

That makes the error a little more useful when glancing through the Problems window without having to click on the items.